[autograd] lower MAX_DEPTH limit according to TSAN limit

wanchaol opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* **#36745 [autograd] lower MAX_DEPTH limit according to TSAN limit**

As we hold a mutex for our custom C++ Node, when calling reentrant
backward from custom C++ function, we will cocurrently holding many
mutexes up to MAX_DEPTH. TSAN only allow 65 mutexes at once, otherwise
it will complain. This PR lower the limit according to TSAN.

TSAN Reference: https://github.com/google/sanitizers/issues/950

Differential Revision: [D21072604](https://our.internmc.facebook.com/intern/diff/D21072604)

wanchaol committed:
[autograd] lower MAX_DEPTH limit according to TSAN limit

As we hold a mutex for our custom C++ Node, when calling reentrant
backward from custom C++ function, we will cocurrently holding many
mutexes up to MAX_DEPTH. TSAN only allow 65 mutexes at once, otherwise
it will complain. This PR lower the limit according to TSAN.

TSAN Reference: https://github.com/google/sanitizers/issues/95

[ghstack-poisoned]

albanD reviewed (approved):
LGTM, thanks for the PR.
Let's make sure it fixes the tsan test before merging.

yf225 commented on torch/csrc/autograd/engine.h:
```diff
@@ -37,6 +37,14 @@ void validate_outputs(
+// As we hold mutex in every of our custom C++ autograd Node, we would
+// like to avoid TSAN complains on this when doing reentrant backwards
+// For reference, see https://github.com/google/sanitizers/issues/950
+static constexpr int MAX_DEPTH = 60;
```
Should this value be different based on whether we have TSAN enabled?

albanD replied:
This should not have any meaningful impact on perf. So I would lean toward consistency between all the builds :D

wanchaol committed:
Update on "[autograd] lower MAX_DEPTH limit according to TSAN limit"


As we hold a mutex for our custom C++ Node, when calling reentrant
backward from custom C++ function, we will cocurrently holding many
mutexes up to MAX_DEPTH. TSAN only allow 65 mutexes at once, otherwise
it will complain. This PR lower the limit according to TSAN.

TSAN Reference: https://github.com/google/sanitizers/issues/950

[ghstack-poisoned]
